File: TUnit.TestProject/DynamicallyRegisteredTests.cs
using System.Diagnostics.CodeAnalysis;
using TUnit.Core;
using TUnit.Core.Interfaces;

namespace TUnit.TestProject;

[DynamicCodeOnly]
public class DynamicallyRegisteredTests
{
    [Test]
    [DynamicDataGenerator]
    public void MyTest(int value)
    {
        throw new Exception($@"Value {value} !");
    }
}

public class DynamicDataGenerator : DataSourceGeneratorAttribute<int>, ITestStartEventReceiver, ITestEndEventReceiver
{
    private static int _count;

    private readonly CancellationTokenSource _cancellationTokenSource = new();

    protected override IEnumerable<Func<int>> GenerateDataSources(DataGeneratorMetadata dataGeneratorMetadata)
    {
        yield return () => new Random().Next();
    }

    public ValueTask OnTestStart(TestContext testContext)
    {
        if (!IsReregisteredTest(testContext))
        {
            testContext.AddLinkedCancellationToken(_cancellationTokenSource.Token);
        }

        return default(ValueTask);
    }

    [Experimental("WIP")]
    [UnconditionalSuppressMessage("Trimming", "IL2026:Members annotated with 'RequiresUnreferencedCodeAttribute' require dynamic access otherwise can break functionality when trimming application code", Justification = "Dynamic Code Only attribute on test")]
    public async ValueTask OnTestEnd(TestContext testContext)
    {
        if (testContext.Result?.State == TestState.Failed)
        {
            await _cancellationTokenSource.CancelAsync();

            // We need a condition to end execution at some point otherwise we could go forever recursively
            if (Interlocked.Increment(ref _count) > 5)
            {
                throw new Exception();
            }

            if (IsReregisteredTest(testContext))
            {
                // Optional to reduce noise
                // testContext.SuppressReportingResult();
            }

            await testContext.ReregisterTestWithArguments(methodArguments: [new Random().Next()],
                objectBag: new()
                {
                    ["DynamicDataGeneratorRetry"] = true
                });
        }
    }

    private static bool IsReregisteredTest(TestContext testContext)
    {
        return testContext.ObjectBag.ContainsKey("DynamicDataGeneratorRetry");
    }

    public int Order => 0;
}

File: TUnit.TestProject.FSharp/DynamicallyRegisteredTests.fs
namespace TUnit.TestProject.FSharp

#nowarn "57"
open System
open System.Collections.Generic
open System.Threading
open System.Threading.Tasks
open TUnit.Core
open TUnit.Core.Enums
open TUnit.Core.Interfaces
open TUnit.Engine.Extensions

type DynamicDataGenerator() =
    inherit DataSourceGeneratorAttribute<int>()
    let mutable count = 0
    let cts = new CancellationTokenSource()

    override _.GenerateDataSources(dataGeneratorMetadata) =
        seq { yield Func<int>(fun () -> (Random()).Next()) }

    interface ITestStartEventReceiver with
        member _.OnTestStart(beforeTestContext: BeforeTestContext) =
            if not (DynamicDataGenerator.IsReregisteredTest(beforeTestContext.TestContext)) then
                beforeTestContext.AddLinkedCancellationToken(cts.Token)
            ValueTask()

    interface ITestEndEventReceiver with
        member _.OnTestEnd(afterTestContext: AfterTestContext) =
            task {
                let testContext: TestContext | null = afterTestContext.TestContext
                match testContext with
                | null -> ()
                | ctx ->
                    if ctx.Result <> null && ctx.Result.Status = Status.Failed then
                        cts.Cancel()
                        count <- count + 1
                        if count > 5 then
                            raise (Exception())
                        if DynamicDataGenerator.IsReregisteredTest(ctx) then
                            () // Optionally suppress reporting
                        let retryDict = Dictionary<string, obj | null>()
                        retryDict.Add("DynamicDataGeneratorRetry", box true)
                        do! ctx.ReregisterTestWithArguments([|(Random()).Next()|], retryDict)
            } |> ValueTask

    interface IEventReceiver with
        member _.Order = 0

    static member private IsReregisteredTest(testContext: TestContext) =
        testContext.ObjectBag.ContainsKey("DynamicDataGeneratorRetry")

[<DynamicCodeOnly>]
type DynamicallyRegisteredTests() =
    [<Test>]
    [<DynamicDataGenerator>]
    member _.MyTest(value: int) =
        raise (Exception($"Value {value} !"))
